subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/VariantResolvingArtifactSet.java

```diff
@@ -0,0 +1,182 @@
+            }
+
+            ResolvedVariantSet variants;
+            try {
+                variants = getVariants(selectFromAllVariants);
+            } catch (Exception e) {
+                return new BrokenResolvedArtifactSet(e);
+            }
+            return selector.select(variants, this);
+        }
+    }
+
+    @Override
+    public ResolvedArtifactSet asTransformed(ResolvedVariant sourceVariant, VariantDefinition variantDefinition, TransformUpstreamDependenciesResolverFactory dependenciesResolverFactory, TransformedVariantFactory transformedVariantFactory) {
+        if (componentId instanceof ProjectComponentIdentifier) {
+            return transformedVariantFactory.transformedProjectArtifacts(componentId, sourceVariant, variantDefinition, dependenciesResolverFactory);
+        } else {
+            return transformedVariantFactory.transformedExternalArtifacts(componentId, sourceVariant, variantDefinition, dependenciesResolverFactory);
+        }
+    }
+
+    public ResolvedVariantSet getVariants(boolean selectFromAllVariants) {
+        ImmutableSet<ResolvedVariant> variants;
+        if (!selectFromAllVariants) {
+            variants = ownArtifacts.get();
+        } else {
+            variants = getComponentVariants();
+        }
+
+        return new DefaultResolvedVariantSet(
+            componentId,
+            schema,
+            overriddenAttributes,
+            variants
+        );
+    }
+
+    public ImmutableSet<ResolvedVariant> calculateOwnArtifacts() {
+        VariantArtifactResolveState variantState = variant.prepareForArtifactResolution();
+
+        if (artifacts.isEmpty()) {
+            ComponentArtifactResolveMetadata componentMetadata = component.prepareForArtifactResolution().getResolveMetadata();
+
+            ImmutableSet.Builder<ResolvedVariant> ownArtifacts = ImmutableSet.builder();
+            visitResolvedArtifacts(componentMetadata, variantState, ownArtifacts::add, exclusions.mayExcludeArtifacts());
+            return ownArtifacts.build();
+        } else {
+            return ImmutableSet.of(variantState.resolveAdhocVariant(variantResolver, artifacts));
+        }
+    }
+
+    /**
+     * Gets all artifact variants ("sub-variants") for the component. This is used when
+     * artifact view variant-reselection is enabled.
+     *
+     * TODO: Currently, this contains all variants in the entire component,
+     * however in practice when using withVariantReselection the user likely
+     * does not want to select from variants with a different capability than
+     * the current variant.
+     */
+    private ImmutableSet<ResolvedVariant> getComponentVariants() {
+        ComponentArtifactResolveState componentState = component.prepareForArtifactResolution();
+        List<VariantArtifactResolveState> componentVariants = componentState.getVariantsForArtifactSelection().orElse(null);
+
+        if (componentVariants == null) {
+            return ownArtifacts.get();
+        }
```

Comment:
🤔 IIUC, this is equivalent:
```suggestion
        List<VariantArtifactResolveState> componentVariants = componentState.getVariantsForArtifactSelection().orElseGet(ownArtifacts);
```

Reply:
I don't think this is. Your version does not return early with the result of `ownArtifacts`.

Reply:
doh, I somehow read this as `componentVariants = ownArtifacts.get()` and not an early return.